Add data source and resource for roles

gonzolino opened the pull request:
Adds `openstack_identity_role_v3` data source & resource.

See #248 for more information (this PR contains part of the content that was initially proposed there).

Docs are still missing.

jtopjian commented:
@gonzolino Thank you for the work you've put in with these resources :smile: 

I haven't done a review yet, but I noticed that the builds are failing. This is because the CI environment doesn't run as an "admin" user, so resources which require admin-level privileges need to be ignored (for now. that should be changed at some point).

There's an example of how to do that here:

https://github.com/terraform-providers/terraform-provider-openstack/blob/master/openstack/resource_openstack_identity_user_v3_test.go#L22-L25

Because the acceptance tests will be ignored, I'll run the tests manually when they're ready for review.

Let me know if you have any questions or need any help :)

gonzolino commented:
@jtopjian Thanks for the hint with the admin only acceptance tests! I updated the patch, so the test should be ignored now.

The tests I added for the role resource & data source are rather simple at the moment. It would be nice if you could have a look at them in your review, and if you have ideas for additional tests just let me know :)

jtopjian commented on openstack/data_source_openstack_identity_role_v3_test.go:
```diff
@@ -0,0 +1,50 @@
+
+const testAccOpenStackIdentityV3RoleDataSource_basic = `
+data "openstack_identity_role_v3" "role_1" {
+    name = "Member"
```
Let's change this to `admin`. Sounds silly, but `Member` might not exist in a testing environment. For example, mine has `_member_`. `admin`, however, is definitely a standard name.

jtopjian commented on openstack/resource_openstack_identity_role_v3.go:
```diff
@@ -0,0 +1,118 @@
+		Read:   resourceIdentityRoleV3Read,
+		Update: resourceIdentityRoleV3Update,
+		Delete: resourceIdentityRoleV3Delete,
+		Importer: &schema.ResourceImporter{
```
In order to ensure `import` works correctly, can you add an import test? See the `import_*_test.go` files as an example.

jtopjian commented on openstack/resource_openstack_identity_role_v3.go:
```diff
@@ -0,0 +1,118 @@
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
```
We'll need to add the `region` attribute here. Just copy and paste it from another resource -- they're all the same configuration.

jtopjian commented on openstack/data_source_openstack_identity_role_v3.go:
```diff
@@ -0,0 +1,81 @@
+	return &schema.Resource{
+		Read: dataSourceIdentityRoleV3Read,
+
+		Schema: map[string]*schema.Schema{
```
We'll need to add the region attribute here. Just copy and paste it from another resource -- they're all the same configuration.

jtopjian commented on openstack/resource_openstack_identity_role_v3_test.go:
```diff
@@ -0,0 +1,96 @@
+	"github.com/gophercloud/gophercloud/openstack/identity/v3/roles"
+)
+
+func TestAccIdentityV3Role_basic(t *testing.T) {
```
Can you add a second step here which updates the name?

gonzolino commented:
@jtopjian Thanks for your comments!

I tried to address all of them and it would be nice if you could have another look.

jtopjian commented:
@gonzolino I apologize for the late reply.

This looks really good. I tested it and it works well. It'll need to be rebased with master -- there was a PR merged the other day which clashes with `openstack/provider.go`. Let me know if you need help with this.

Once the merge conflict is resolved and docs are added, this is good to go. :)

gonzolino commented:
@jtopjian No worries :)
Rebase is done and docs are added. So the PR is ready for the next review round 👍

jtopjian commented on openstack/resource_openstack_identity_role_v3.go:
```diff
@@ -0,0 +1,125 @@
+
+	role, err := roles.Get(identityClient, d.Id()).Extract()
+	if err != nil {
+		return CheckDeleted(d, err, "project")
```
`s/project/role`

jtopjian commented on openstack/resource_openstack_identity_role_v3.go:
```diff
@@ -0,0 +1,125 @@
+		}
+	}
+
+	return resourceIdentityProjectV3Read(d, meta)
```
`s/resourceIdentityProjectV3Read/resourceIdentityRoleV3Read`

The Update step of the acceptance test was failing because of this - took me a while to spot it :smile:

jtopjian commented:
recheck

gonzolino commented:
@jtopjian Sorry for taking so long, I hope all your comments are addressed now :)

jtopjian commented:
@gonzolino Nice work! This looks good to me.
